### PR TITLE
Improved algorithm for index 'dry_spell_total_length' with new option…

### PR DESCRIPTION
**Pull request by yrouranos**
…s (operator 'max', possibility to specify start/end dates) and calculation for days near dataset boundaries.

### Pull Request Checklist:
- [x] This PR addresses an already opened issue (for bug fixes / features)
    - This PR adds features and fixes problems described in #843
- [x] Tests for the changes have been added (for bug fixes / features)
    - 1 test has been added in test_precip.py.
    - 1 test has been added in test_indices.py.
    - 30 additional tests have been done outside of xclim to test the algorithm in detail (see scen_workflow_afr.unit_tests).
- [x] Documentation has been added / updated (for bug fixes / features)
- [x] HISTORY.rst has been updated (with summary of main changes)
    - [ ] Link to issue (:issue:`number`) and pull request (:pull:`number`) has been added.
- [ ] `bumpversion patch` has been called on this branch
- [x] The relevant author information has been added to `.zenodo.json`

### What kind of change does this PR introduce?
* A new argument allows to specify the resampling operator. The initial version of the algorithm was resampling according to the `sum` operator. The improved algorithm gives the possibility to resample using the `max` operator, which means that a sequence of ` window` dry days can happen if there is a single day with precipitation reaching or exceeding a threshold (`tresh`).
* It is now possible to specify a `start_date` and `end_date`. The days comprised in a sequence of `window` dry days are first identified. A mask is then applied to this result using the provided start and end dates. Resampling at the requested frequency (`freq`) is finally done. The implication of this change is that it is possible to have fewer than `window` days in a year, because some of these days could have been removed by the mask.
* The algorithm allows a `start_date` (ex: "12-01") to be larger than `end_date` (ex: "02-28"). In this example, the dry days between December and February are not considered.
* Calculation will be correct when specifying a `window` size that is an even number. The previous version of the algorithm was pushing a sequence of  dry days to the right by one day with an even `window` size, assigning a wet label to the first day of a dry sequence, and a dry label to the day following the sequence. It can be an issue if one dry day is transferred to the following year (when a dry period overlaps two years). This situation seems to have been caused by the centre-based window, which is asymmetric if it's size is an even number.

### Does this PR introduce a breaking change?
* With the previous algorithm, the fist `window / 2` days and last `(window - 1) / 2` days of the input precipitation dataset were assigned `nan` value, and thus were assumed to be wet (because they were not dry), which will result in unrealistic indices in certain climates, e.g., West Africa (there is no precipitation between December and February). Ideally, the first and last years of the computed indices should be excluded, but the user may be unaware of this situation, which implies a risk of misusing the function. The combination of a left-based window and a bidirectional scanning (start to end of the input dataset, and the opposite), allows to calculate indices that consider the first and last days, without risk.

### Other information:
* It is possible that climate indices behave in a similar way as the current `dry_spell_total_length` algorithm (indices based on a `window`, centre-based or not). A similar solution, or an improved one, could be implemented for other indices.
* This is certainly the case with `dry_spell_frequency`, which should work exactly the same way as `dry_spell_total_length`.

### CODE REVIEW

**Comment by yrouranos** [comment]
@aulemahal I updated the algorithm to account for bisextile years. I removed xr.DataArray casting. I also added 2 test cases (see test_indices.py:test_dry_spell).

File: xclim/indices/_agro.py [hunk 1]
```diff
@@ -628,11 +629,17 @@ def dry_spell_frequency(
...
+    op: str = "sum",
+    freq: str = "YS",
+    start_date: str = "",
+    end_date: str = "",
 ) -> xarray.DataArray:
     """
-    Return the total number of days in dry periods of n days and more, during which the accumulated precipitation
-    on a window of n days is under the threshold.
+    Return the total number of days in dry periods of n days and more, during which the maximum or accumulated
+    precipitation on a window of n days is under the threshold.
```
Comment (huard) [inline xclim/indices/_agro.py:642]: I suggest removing the reference to *n* in the title, and add a paragraph below to include details on the algorithm used.

Comment (huard) [reply xclim/indices/_agro.py:642]: Also add references to papers or reports defining or using this algo.

**Review by huard — CHANGES_REQUESTED** [review]

**Comment by aulemahal** [comment]
@yrouranos  Comme je n'ai pas accès en écriture à ta branche, j'ai fait une 3e branche que je veux fusionner dans celle-ci pour ensuite fusionner dans le master. Ma PR est donc ouverte dans ton repo.
On peut aussi faire autrement, si tu veux : je pourrais fermer cette PR et la remplacer par une nouvelle à partir de ma tierce branche.

**Comment by aulemahal** [comment]
@yrouranos  J'ai fusionné mes changements et le master! J'ai aussi ajusté la docstring et ajouté une entrée au "history".

@huard J'ai ajouté une note pour soulever le problème qui survient aux bords. Je ne pense pas par contre qu'il soit nécessaire de mettre une référence ici? Ça reste un calcul simple?

**Comment by yrouranos** [comment]
J'ai 2 commentaires et une suggestion:
1. À propos de la note, je crois qu'on assume plutôt que les jours précédant ou suivant la série sont secs. S'il pleut le jour no. 1 de la série, les jours qui précèdent (pour lesquels les données ne sont pas disponibles) n'ont aucune importance étant donné que nous savons que le jour no. 1 ne fait pas partie d'une séquence de jours secs étant donné qu'il a plu cette journée; considérer les jours qui précèdent la série ne changerait rien. Par contre, si le jour no. 1 est sec, il pourrait ne pas faire partie d'une séquence de jours secs en réalité s'il avait plu lors du jour précédent le jour no. 1. Dans ce cas, ça voudrait dire que la condition est moins stricte près des frontières (nous avons besoin de 3 jours consécutifs près des frontières, au lieu de 5 ailleurs).
2. C'est une bonne idée d'avoir ajouté une note étant donné que ça laisse la possibilité à l'utilisateur d'exclure la première et dernière année si l'hypothèse est problématique pour le contexte d'une étude. Pour cet indice, je crois que l'hypothèse sera rarement problématique. Par contre, si un indice similaire visait à calculer la période sèche la plus longue par année, un utilisateur intéressé au climat d'une région de l'hémisphère sud (ex: Afrique de l'Ouest) pourrait avoir à exclure la première et la dernière année (la période sèche débute en novembre et se termine au mois de mars). Les indices calculés n'auraient pas de sens pour les années 1 et n.
3. J'ai essayé le nouvel algorithme, et les cas ont tous fonctionné, sauf les no. 20-21. J'ai réalisé que c'était causé par une date de début non spécifiée (cas no. 20) ou une date de fin non spécifiée (cas no. 21). L'algorithme que j'avais développé, assumais que le premier jour était le 1er janvier (si non spécifié) et que le dernier jour était le 31 décembre (si non spécifié). Je me demandais si ça vaudrait la peine d'intégrer ce comportement dans select_time(), même si ce n'est pas indispensable. L'utilisateur spécifiera normalement les deux dates.

**Comment by aulemahal** [comment]
1. J'ai l'impression que ma note est plus proche du code. Dans mon exemple, `window = 3`. Donc les "périodes sèches" sont par définition de longueur 3. Une période plus longue est simplement une superposition de périodes de longueur 3. Le jour 0 (allons-y à la numérotation python) fait partie de 3 périodes candidates : (-2, -1, 0), (-1, 0, 1) et (0, 1, 2). 

Ligne 698 : Il n'y pas de données pour -2 et -1, et le drapeau booléen de "période sèche" est attribué au dernier jour de la période, donc 0 et 1 sont NaN.
Ligne 699 : On refait un rolling pour voir si l'une de ces 3 périodes est sèche (et donc le jour fait partie d'une période sèche). Comme on fait un shift de 2 vers la gauche, c'est la valeur à l'indice 2 qui ramenée au jour 0. On aura True, si on a au moins un True dans 0, 1, 2. Comme 0 = 1 = NaN, il faut absolument que (0, 1, 2) soit une période sèche. Ce qui insinue que  (-2, -1, 0) et (-1, 0, 1) ne sont pas sèches, et donc c'est pourquoi je dis que nous assumons implicitement que -2 et -1 sont mouillés.

3. En effet, `select_time` s'attend à ce que les deux bornes soient données. Comme la fonction ne prend pas  `freq` en entrée, donc elle ne peut pas savoir qu'elle date/doy serait le plus logique à mettre. Mais ça pourrait être modifié, cette information existe.

File: xclim/testing/tests/test_indices.py [hunk 2]
```diff
@@ -2471,91 +2471,123 @@ def test_water_budget(pr_series, tasmin_series, tasmax_series):
...
+            thresh, window = 3, 7
+            out_events = [2]
+            out_total_d_sum = [18]
+            out_total_d_max = [20]
+            pr = pr_2
+        else:
+            thresh, window = 1, 14
+            out_events = [0]
+            out_total_d_sum = out_total_d_max = [7, 7]
+            pr = pr_3
```
Comment (huard) [inline xclim/testing/tests/test_indices.py:2517]: ce n'est pas le genre de trucs qu'on paramétrise généralement avec pytest ?

**Review by huard — COMMENTED** [review]
J'ai un peu l'impression que les effets de bords de la méthode sont indésirables, et non un feature. 

Il n'y a pas d'indicator associé ?

File: xclim/indices/_agro.py [hunk 3]
```diff
@@ -660,21 +666,40 @@ def dry_spell_total_length(
...
+
+    Notes
+    -----
+    The algorithm assumes days before and after the timeseries are "wet", meaning that
+    the condition for being considered part of a dry spell is stricter on the edges. For
+    example, with `window=3` and `op='sum'`, the first day of the series is considered
+    part of a dry spell only if the accumulated precipitation within the first 3 days is
+    under the threshold. In comparison, a day in the middle of the series is considered
+    part of a dry spell if any of the three 3-day periods of which it is part are
+    considered dry (so a total of five days are included in the computation, compared to only 3.)
```
[the remark this answers is not in the document]
Comment (aulemahal) [reply xclim/indices/_agro.py:692]: Que proposes-tu à la place? Des NaNs à chaque bout? Genre, toute l'année NaN parce qu'il nous manque 2 jours?

[the remark this answers is not in the document]
Comment (huard) [reply xclim/indices/_agro.py:692]: Je me questionne plutôt sur le sens "climatologique" de l'indice, pas sur la technique de calcul.

[the remark this answers is not in the document]
Comment (aulemahal) [reply xclim/indices/_agro.py:692]: Je ne comprends toujours pas.
Le choix qui est fait ici a été de calculer la _séchitude_ des jours aux bords malgré un manque de données. Dans l'itération précédente, un NaN était assigné, tout simplement. Le nouvel algorithme n'a pas de faux positif, et dans le pire cas, `window - 1` faux négatifs de chaque bord. 
Je note que dans la majorité des cas, l'utilisateur devrait être au courant de la climatologie de ses données et donc savoir à quel moment de l'année les périodes sèches pourraient arriver. Ainsi, il choisi `freq` pour faire tomber la frontière entre périodes dans la période la plus mouillée possibe.

[the remark this answers is not in the document]
Comment (huard) [reply xclim/indices/_agro.py:692]: Je comprend mieux. En gros, le comportement du "milieu" est le comportement souhaité, et il est un peu différent sur les bords mais c'est préférable à un NaN ?

File: xclim/testing/tests/test_indices.py (hunk 2 above)
Comment (aulemahal) [reply xclim/testing/tests/test_indices.py]: Fait. Je trouve que c'en est un peu moins facile à lire, mais en effet mainteant c'est cohérent avec le reste.

File: xclim/indices/_agro.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (aulemahal) [reply xclim/indices/_agro.py:693]: Je pense oui. Et "les bords" ce sont `window - 1` jours sur 365 dans la majorité des cas, ça reste peu.

[the remark this answers is not in the document]
Comment (yrouranos) [reply xclim/indices/_agro.py:693]: Avec l'algorithe précédent, la présence de NA dans le contexte du climat Ouest-Africain était problématique. Attribuer un NA revenait à dire (indirectement) que les `plafond((windows - 1) / 2)` jours de part et d'autre de la frontière étaient pluvieux étant donné que ces jours n'étaient pas comptabilité dans les jours secs (seulement 2 possibilités, même en mettant un NA). Et l'usager n'était pas averti non plus. À mon avis, l'avertissement dans le header de la fonction est utile/important et suffisant. Et oui, les usagers sont habituellements au courant de leur climatologie, donc si l'effet de bordure est problématique (par exemple, si l'hypothèse d'une journée sèche le 31 décembre ne tient pas la route), alors ils peuvent ignorer les indices calculés pour la première et la dernière année, donc considérer les indices de 28/30 ans.

Je ne suis pas certain comment on pourrait compllètement éliminer l'hypothèse près des bordures. Avec le nouvel algorithme, l'appartenance d'une jounée à une séquence sèche est quand même vérifiée, mais d'un seul côté. J'avais également pensé à ajouter un nouveau paramètre que l'utilisateur peut utiliser pour dire si les journées près de la frontière sont habituellement sèches ou pluvieuses. Dans le cas de l'Afrique de l'Ouest, j'indiquerais qu'on s'attend à des jours secs. À Montréal, la probabilité serait peut-être plus importante d'avoir une journée pluvieuse. Et dans d'autres contextes, j'imagine qu'il n'y aurait pas vraiment de tendance claire, ce qui veut dire que l'usager se retrouverait dans la même situation qu'avec l'algorithme proposé.

À propos de l'assignation automatique d'une date manquante (ex : `start_date` à partir d'une `end_date`), je voyais 2 options. La 1ière option est celle que j'expliquais précédamment, donc choisir le 1er ou dernier jour de l'année, selon l'information manquante. La 2ième option est de prendre le jour précédent de l'année suivante (ou prendre le jour suivant de l'année précédente). Par exemple, si la `start_date` fournie est le 3 mars 1987, alors la `end_date` (devinée) pourrait être le 2 mars 1988. Je me posais la question parce qu'un usager peut spécifier ou pas les dates. Si les dates ne sont pas spécifiées et que la fréquence des résultats est annuelle, alors on prend quand même la décision que les 1er et derniers jours sont le 1er janvier et le 31 décembre (selon ma compréhension).

[the remark this answers is not in the document]
Comment (huard) [reply xclim/indices/_agro.py:693]: ok merci, bon pour moi.

[the remark this answers is not in the document]
Comment (aulemahal) [reply xclim/indices/_agro.py:693]: Pour les dates, j'avoue que je suis principalement un peu paresseux ici.
`select_time` c'est pour sélectionner un portion de l'année, donc la 2e option n'est pas cohérente avec la version actuelle de la fonction et de son utilisation.
Pour faire la 1re, il faut que xclim sache l'année commence et se termine quand. Pour le moment `select_time` ne reçoit pas cette information, mais ce serait facile à faire, suffit de passer `freq` (qui est "YS" ou "AS-JUL", ou qqch d'autre d'annuel).
Par contre, l'utilisateur donne déjà `freq` à la fonction, donc il est prescrit déjà un découpage des années. Ma paresse vient du fait que je ne vois pas de cas où l'utilisateur ne connaît pas l'une des deux bornes, mais que xclim oui.

[the remark this answers is not in the document]
Comment (yrouranos) [reply xclim/indices/_agro.py:693]: Effectivement, les usagers connaissent habituellement les deux dates. Donc, s'il manque une date, ça serait probablement dû à une erreur d'inattention de la part de l'utilisateur.

**Review by aulemahal — APPROVED** [review]

**Pull request merged**